lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -173,243 +144,354 @@ class ExpandWhensVisitor : public FIRRTLVisitor<ExpandWhensVisitor> {
     return newConnect;
   }
 
-private:
-  /// Map of destinations and the operation which is driving a value to it in
-  /// the current scope. This is used for resolving last connect semantics, and
-  /// for retrieving the responsible connect operation.
-  llvm::MapVector<FieldRef, Operation *> &scope;
+  void visitDecl(WireOp op) { declareSinks(op.result(), Flow::Duplex); }
 
-  /// The current wrapping condition. If null, we are in the outer scope.
-  Value condition;
-};
-} // namespace
-
-LogicalResult ExpandWhensVisitor::run(FModuleOp module) {
-  llvm::MapVector<FieldRef, Operation *> outerScope;
-  ExpandWhensVisitor(outerScope, nullptr).visitDecl(module);
-
-  // Check for any incomplete initialization.
-  LogicalResult result = success();
-  for (auto destAndConnect : outerScope) {
-    // If there is valid connection to this destination, everything is good.
-    auto *connect = std::get<1>(destAndConnect);
-    if (connect)
-      continue;
+  void visitDecl(RegOp op) {
+    // Registers are initialized to themselves.
+    // TODO: register of bundle type are not supported.
+    assert(!op.result().getType().isa<BundleType>() &&
+           "registers can't be bundle type");
```

Comment:
This message could be changed to `ExpandWhens does not support BundleType (this should have been removed by LowerTypes)` or something. Also, what about `VectorType`?

Reply:
Hmm  I have to think about this.  Most of the pass *does* support bundle types right now, because we have considered moving this pass before lower types. The part missing for Regs is creating the self connect - It has to recursively create a self connection for each field. The pass as a whole is missing vector support since FieldRefs weren't working for vector types yet.  Neither of these things would be too hard.

BundleTypes can show up from memories after lower types, and we need to make sure those are initialized correctly.   If we could guarantee that there were no BundleTypes here, then we could regress the pass functionality and it would probably run a bit quicker.

Reply:
I just deleted the assert.  It was sort of more a written reminder that the functionality had to be implemented, but there is a TODO right there.  Its could only have triggered if we tried moving this pass before lowertypes and watched what happened.